torch/distributed/fsdp/wrap.py

```diff
@@ -209,6 +211,48 @@ def __repr__(self) -> str:
         return super().__repr__() + f"({self._module_classes_str})"
 
 
+class LambdaWrapPolicy(_WrapPolicy):
+    """
+    This policy takes in a lambda function that maps a given ``nn.Module`` to
+    either ``False``, ``True``, or a FSDP kwarg dictionary.
+    - If the function returns ``False`` or an empty dictionary, then the module
+      is not wrapped.
+    - If the function returns ``True``, then the module is wrapped using the
+      root's kwargs.
+    - If the function returns a non-empty dictionary, then the module is
+      wrapped, and the dictionary overrides the root's kwargs.
+    """
+
+    def __init__(self, lambda_fn: Callable[[nn.Module], Union[bool, Dict[str, Any]]]):
+        self._lambda_fn = lambda_fn
+
+    def _run_policy(
+        self,
+        root_module: nn.Module,
```

Comment:
OOC, is there a way to easily determine the FQN of the module at this point? Without the root module it seems like it might be difficult, whereas, the code that calls `_run_policy` would be able to generate it without much difficulty.

Reply:
That is a good point. I do think we need access to the root module to be able to generate FQNs ("we" encompasses both in FSDP and for users).

Reply:
In a downstream package I maintain, I've been allowing users to [provide module FQN-based complements](https://finetuning-scheduler.readthedocs.io/en/latest/advanced/fsdp_scheduled_fine_tuning.html#:~:text=users%20to%20provide-,module%20name%2Dbased%20complements,-to%20a%20given) to a given ``auto_wrap_policy`` using a custom ``_FSDPPolicy`` ([NameDrivenPolicy](https://github.com/speediedan/finetuning-scheduler/blob/3b70bbe0e00347ea00ea52c92ea47804193f0640/src/finetuning_scheduler/strategy_adapters/fsdp.py#L843C1-L866C113)). I should be able to continue offering that functionality with some minor refactoring after this PR, but I thought it might be worth mentioning in the context of this PR because: 
1. To call attention to this pattern since I've received positive user feedback regarding the utility FQN-based wrapping (to accommodate more flexible fine-tuning requirements in this case)
2. In case the team considers the functionality sufficiently useful to offer it natively at some point in the future (I'm not necessarily advocating that, just thinking aloud... either way this PR should help make such extensions cleaner)

Anyway, nice enhancement @awgu! A pleasure as always working with this thoughtfully designed API and constantly improving package.

Reply:
Thanks for the pointer! This is super helpful, especially learning more about the customer feedback.

I am thinking that `CustomPolicy` can stay generic, and we can push specific logic to the `lambda_fn` construction. If the user wants to use FQNs, then the user can use the same trick as your implementation where you translate FQNs to Python `id()`s and then use the `id()`s in the `lambda_fn`, which will see the `module` as argument. Like you said, after some refactoring, you should still be able to make use of `CustomPolicy`. The question is whether we should offer some API that permits FQNs directly. I think that is something that we can wait and see. For now, we can just add `CustomPolicy` that can serve as a base for other extensions.

What do you think?

Reply:
> The question is whether we should offer some API that permits FQNs directly. I think that is something that we can wait and see. For now, we can just add `CustomPolicy` that can serve as a base for other extensions.

Agree completely, I think prioritizing API stability and simplicity at this juncture is the prudent approach. Adding an API that permits FQNs directly might be worth considering at some point in the future but that feature isn't sufficiently crucial to justify its inclusion at present (especially since implementation approaches like my `id()` translation one we discussed aren't too onerous).

Thanks for the quick and thoughtful response!